include/pybind11/cast.h

```diff
@@ -1042,12 +1042,31 @@ struct type_caster<T, enable_if_t<std::is_arithmetic<T>::value && !is_std_char_t
             return false;
         } else if (!convert && !index_check(src.ptr()) && !PYBIND11_LONG_CHECK(src.ptr())) {
             return false;
-        } else if (std::is_unsigned<py_type>::value) {
-            py_value = as_unsigned<py_type>(src.ptr());
-        } else { // signed integer:
-            py_value = sizeof(T) <= sizeof(long)
-                ? (py_type) PyLong_AsLong(src.ptr())
-                : (py_type) PYBIND11_LONG_AS_LONGLONG(src.ptr());
+        } else {
+            handle obj = src;
+#if PY_VERSION_HEX < 0x03080000
+            bool do_decref = false;
+            if (index_check(src.ptr())) {
+                PyObject *tmp = PyNumber_Index(src.ptr());
+                if (!tmp) {
+                    PyErr_Clear();
+                    return false;
+                }
+                do_decref = true;
```

Comment:
`index_owner = reinterpret_steal<object>(tmp);`

That way you don't need the second `#if PY_VERSION_HEX < 0x03080000` below and this code become exception safe.

I'd also use `idx` (or similar) instead of `tmp`, to be more descriptive.

Reply:
I thought of/tried that, but didn't want to incur an overhead refcounting on Python >= 3.8, and this is also what CPython does. 

But wait, maybe you mean something else, that doesn't need this! I'll give this a shot :-)

Reply:
Yesss, that does work out beautifully! Thanks!

Reply:
Still seeing if this could easily be refactored out.

Reply:
OK, it's quite hard to refactor into a separate private function without incurring an additional `inc_ref`/`dec_ref`, it seems. It's already cleaner than before, though, so is it fine to leave like this for now?